### ISSUE TITLE
Auto populate ModelFactory::getDefaults() based on doctrine mapping

### ISSUE THREAD

**Issue by MGDSoft** [zenstruck/foundry#100]
Hi @kbond thx for this bundle!

Two years ago I created a similiar bundle https://github.com/MGDSoft/FixturesGeneratorBundle. 
This bundle can create auto defaultValues and create fixtures for all entities with a single command (with his dependencies), it would be great if that features was in this bundle too.
I can help with this if you are agree.

**Comment by kbond** [comment]
Hey @MGDSoft, glad you like the bundle!

#15 I think would address "create fixtures for all entities with a single command" so that one I'd love some help on that.

For the auto default values, this would be really neat. I'm trying to think on how it would work:

1. I'd only like non-nullable values to have a default auto set.
2. For non-relation fields, setting a default value via faker should be easy enough (ie `$faker->sentence()` for `string`, `$faker->randomNumber()` for `integer`, `$faker->dateTime()` for `datetime`, etc...)
3. For non-null relations this might get a bit more difficult as the factory might not exist for the relation (if not using `--all`). What do you think?

**Comment by benblub** [comment]
👀  is this feature in building?

**Comment by kbond** [comment]
Not yet by me anyway - my schedule's pretty full currently.

**Comment by benblub** [comment]
mhh i like this idea :/

some dummy/inspire code/test
```php
$classMetaData = $entityManager->getClassMetadata(Product::class);

$defaultsArray = [];
$fakerTypes = [
    'integer' => faker()->randomNumber(),
    'string' => faker()->sentence(),
    'date' => faker()->dateTime(),
];

foreach ($classMetaData->fieldMappings as $property) {
    // create with faker
    if (!$property['nullable']) {
        $defaultsArray[] = $fakerTypes[$property['type']];
    }
}

foreach ($classMetaData->associationMappings as $classname => $values) {
    // Check Factory exist and create or skip? class_exists($classname.'Factory')
}
```

```php
/**
 * @ORM\Entity
 * @ORM\Table(name="products")
 */
class Product
{
    /**
     * @ORM\Id
     * @ORM\Column(type="integer")
     * @ORM\GeneratedValue
     */
    protected int $id;
    /**
     * @ORM\Column(type="string", nullable=true)
     */
    protected string $name;

    /**
     * @ORM\Column(type="date")
     */
    protected DateTime $someDate;

    /**
     * @ORM\Column(type="array")
     */
    protected array $someArray = [];

    /**
     * @ManyToOne(targetEntity="Category")
     * @JoinColumn(name="category_id", referencedColumnName="id")
     */
    protected Category $category;


    public function getId()
    {
        return $this->id;
    }

    public function getName()
    {
        return $this->name;
    }

    public function setName($name)
    {
        $this->name = $name;
    }
```

**Comment by kbond** [comment]
Thanks, yes, that's basically what I was thinking. Pretty easy for the scalar doctrine types. It gets more complicated for the non-nullable `ManyToOne` relationships. #162 allows us to now see if a factory exists already so I think auto-creating if it doesn't exist is good idea.

I think this should be an opt-in flag on the maker command (ie `make:factory --auto-defaults`).

**Comment by benblub** [comment]
Its need take care of Asserts too. As examble a string $email with an Assert/Email must use faker()->email() instead of faker()->sentence().

i have to investigate on relationships and yes auto-creating for non exist sounds like a good solution.

```php
    public function getPropertiesFromDoctrineFieldMappings(ReflectionClass $entity)
    {
        $reader = new \Doctrine\Common\Annotations\AnnotationReader();

        $classMetaData = $this->em->getClassMetadata($entity->getName());
        $classAnnotation = $reader->getClassAnnotations($entity); // Not really needed yet

        foreach ($classMetaData->fieldMappings as $property) {
            if (!$property['nullable']
                && in_array($property['type'],  self::DOCTRINE_SCALAR_TYPES)
                && $property['fieldName'] !== 'id') { // TODO get identifier instead of using hardcoded fieldname id

                // TODO we have to read the annotations from props to figure out as examble Assert/Email so that faker generates some email
                // TODO if instance of Constraints or something similar .. maybe add this types to createPropertyDefaultWithFaker
                $reflectionProperty = new ReflectionProperty($entity->getName(), $property['fieldName']);
                $propertyAnnotation = $reader->getPropertyAnnotations($reflectionProperty); // TODO Loop and use Constraints
                $this->defaultProperties[$property['fieldName']] = $this->createPropertyDefaultWithFaker($property['type']);
            }
        }

        return $this->defaultProperties;
    }
```

**Comment by benblub** [comment]
I think this PR can be closed?

**Comment by kbond** [comment]
Closed with #173.
